File: week_10/data_structures/linked_list_examples/swift/LinkedList.swift
struct LinkedList<T: Equatable>: ExpressibleByArrayLiteral {
    private var head: Node<T>

    init() {
        head = .tail
    }

    init(arrayLiteral elements: T...) {
        self.init()
        elements.reversed().forEach { v in
            add(v)
        }
    }

    var count: Int {
        var counter = 0
        var node = head
        while !node.isTail {
            counter += 1
            node = node.next
        }
        return counter
    }

    mutating func add(_ value: T) {
        let oldHead = head
        head = .node(value: value, next: oldHead)
    }

    subscript(_ idx: Int) -> T? {
        guard idx >= 0 else { return nil }

        if idx == 0 { return head.value }

        var searchedNode = head
        for _ in 1...idx {
            searchedNode = searchedNode.next
        }

        if searchedNode.isTail {
            return nil
        } else {
            return searchedNode.value
        }
    }
}

File: week_10/data_structures/linked_list_examples/java/LinkedList.java
public class LinkedList {
  Node head;

  public LinkedList(Node head) {
    this.head = head;
  }

  public int length() {
    int count = 0;
    Node node = head;

    while (node != null) {
      count++;
      node = node.getNext();
    }
    return count;
  }

  public void push(Node newHead) {
    Node existingHead = head;
    this.head = newHead;
    newHead.setNext(existingHead);
  }

  public String get(int index) {
    if (index == 0) {
      return head.getName();
    }
    Node node = head;
    for (int i = 1; i <= index; i++) {
      node = node.getNext();
    }
    return node.getName();
  }
}
